cpp/src/arrow/types/string.cc

```diff
@@ -61,6 +61,15 @@ Status StringArray::Validate() const {
   return BinaryArray::Validate();
 }
 
-TypePtr BinaryBuilder::value_type_ = TypePtr(new UInt8Type());
+// This used to be a static member variable of BinaryBuilder, but it can cause
+// valgrind to report a (spurious?) memory leak when needed in other shared
+// libraries. The problem came up while adding explicit visibility to libarrow
+// and libarrow_parquet
+static TypePtr kBinaryValueType = TypePtr(new UInt8Type());
```

Comment:
So far so good; let's await the visibility patch to get green (still wrestling with builds, if you wouldn't mind a quick review of https://github.com/apache/parquet-cpp/pull/135 it would help me)